src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java

```diff
@@ -0,0 +1,131 @@
+
+  private static final Logger LOGGER =
+    LoggerFactory.getLogger(SchedulerDriverFactory.class);
+
+  /**
+   * If this environment variable is present in the scheduler environment, the master is using
+   * some form of sidechannel auth. When this environment variable is present, we should always
+   * provide a {@link Credential} with (only) the principal set.
+   */
+  private static String SIDECHANNEL_AUTH_ENV_NAME = "DCOS_SERVICE_ACCOUNT_CREDENTIAL";
+
+  /**
+   * Creates and returns a new {@link SchedulerDriver} without a credential secret.
+   *
+   * @param scheduler The Framework {@link Scheduler} implementation which should receive callbacks
+   *     from the {@link SchedulerDriver}
+   * @param frameworkInfo The {@link FrameworkInfo} which describes the framework implementation.
+   *     The 'principal' field MUST be populated and non-empty
+   * @param masterUrl The URL of the currently active Mesos Master, of the form "zk://host/mesos"
+   * @return A {@link SchedulerDriver} configured with the provided info
+   * @throws IllegalArgumentException if {@link FrameworkInfo}.principal is unset or empty when
+   *     authentication is needed
+   */
+  public SchedulerDriver create(
+      final Scheduler scheduler, final FrameworkInfo frameworkInfo, final String masterUrl) {
+    return create(scheduler, frameworkInfo, masterUrl, null /* credentialSecret */);
+  }
+
+  /**
+   * Creates and returns a new {@link SchedulerDriver} with the provided credential secret.
+   *
+   * @param scheduler The Framework {@link Scheduler} implementation which should receive callbacks
+   *     from the {@link SchedulerDriver}
+   * @param frameworkInfo The {@link FrameworkInfo} which describes the framework implementation.
+   *     The 'principal' field MUST be populated and non-empty
+   * @param masterUrl The URL of the currently active Mesos Master, of the form "zk://host/mesos"
+   * @param credentialSecret The secret to be included in the framework
+   *     {@link org.apache.mesos.Protos.Credential}, ignored if {@code null}/empty
+   * @return A {@link SchedulerDriver} configured with the provided info
+   * @throws IllegalArgumentException if {@link FrameworkInfo}.principal is unset or empty when
+   *     authentication is needed
+   */
+  public SchedulerDriver create(
+      final Scheduler scheduler,
+      final FrameworkInfo frameworkInfo,
+      final String masterUrl,
+      final byte[] credentialSecret) {
+    Credential credential;
+    if (credentialSecret != null && credentialSecret.length > 0) {
+      // User has manually provided a Secret. Provide a Credential with Principal + Secret.
+      // (note: we intentionally avoid logging the content of the credential secret, just in case)
+      LOGGER.info("Creating secret authenticated MesosSchedulerDriver for "
+          + "scheduler[{}], frameworkInfo[{}], masterUrl[{}], credentialSecret[{} bytes]",
+          scheduler, frameworkInfo, masterUrl, credentialSecret.length);
```

Comment:
I'd like to see the principal in this log message too, but I guess it's included in the frameworkInfo?